launch_testing/launch_testing/pytest/plugin.py

```diff
@@ -0,0 +1,528 @@
+# Copyright 2021 Open Source Robotics Foundation, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import asyncio
+from collections.abc import Sequence
+import functools
+import inspect
+
+from _pytest.outcomes import fail
+from _pytest.outcomes import skip
+
+import launch
+import launch_testing
+
+import pytest
+
+from .fixture import finalize_launch_service
+from .fixture import get_launch_service_fixture
+
+"""
+launch_testing native pytest based implementation.
+"""
+
+
+try:
+    from _pytest.python import transfer_markers
+except ImportError:  # Pytest 4.1.0 removes the transfer_marker api (#104)
+
+    def transfer_markers(*args, **kwargs):  # noqa
+        """Noop when over pytest 4.1.0"""
+        pass
+
+
+class LaunchTestWarning(pytest.PytestWarning):
+    """Raised in this plugin to warn users."""
+
+    pass
+
+
+def pytest_configure(config):
+    """Inject launch_testing marker documentation."""
+    config.addinivalue_line(
+        'markers',
+        'launch_testing: '
+        'mark the test as a launch test, it will be '
+        'run using the specified launch_pad.',
+    )
+
+
+# TODO(ivanpauno): Deduplicate with launch_testing
+def iterate_ready_to_test_actions(entities):
+    """Search recursively LaunchDescription entities for all ReadyToTest actions."""
+    for entity in entities:
+        if isinstance(entity, launch_testing.actions.ReadyToTest):
+            yield entity
+        yield from iterate_ready_to_test_actions(
+            entity.describe_sub_entities()
+        )
+        for conditional_sub_entity in entity.describe_conditional_sub_entities():
+            yield from iterate_ready_to_test_actions(
+                conditional_sub_entity[1]
+            )
+
+
+def get_ready_to_test_action(launch_description):
+    """Extract the ready to test action from the launch description."""
+    gen = (e for e in iterate_ready_to_test_actions(launch_description.entities))
+    try:
+        ready_action = next(gen)
+    except StopIteration:  # No ReadyToTest action found
+        ready_action = launch_testing.actions.ReadyToTest()
+        launch_description.append(ready_action)
+        return ready_action
+    try:
+        next(gen)
+    except StopIteration:  # Only one ReadyToTest action must be found
+        return ready_action
+    raise RuntimeError(
+        'launch_pytest fixtures must return a LaunchDescription '
+        'containing only one ReadyToTest action'
+    )
```

Comment:
@ivanpauno meta: just like `launch_testing`, this cannot really detect all `ReadyToTest` instances (thinking of includes). We need to document the fact that we'll simply pick up the first one found we can statically find. In which case, I'm slightly inclined to make this a warning. Thoughts?